aesd-char-driver/main.c

```diff
@@ -195,14 +197,109 @@ ssize_t aesd_write(struct file *filp, const char __user *buf, size_t count,
     //goto implementation
 exit_write:
     mutex_unlock(&device->dev_lock);
+    device->written_bytes =retval;
     return retval;
 }
+
+loff_t custom_seek(struct file *file, loff_t offset, int cur_pos)
+{
+    struct aesd_dev *devptr = file->private_data;
+    
+    loff_t mov_cur;
+    switch (cur_pos) 
+    {
+    case 0: /* SEEK_SET */
+        mov_cur = offset;
+        break;
+
+    case 1: /* SEEK_CUR */
+        mov_cur = file->f_pos + offset;
+        break;
+
+    case 2: /* SEEK_END */
+        mov_cur = devptr->written_bytes;
+        break;
+
+    default:
+        return -EINVAL;
+    }
+    file->f_pos = mov_cur;
+    return mov_cur;
+}
+
+long custom_ioctl(struct file *file, unsigned int cmd, unsigned long arg) 
+{
+    long ret=0;
+    struct aesd_dev *device = NULL;
+    struct aesd_seekto buffer;
+    struct aesd_dev *dev = file->private_data;
+    int rc=0,iter=0;
+    long cur_pos;
+    /* Reference - Chapter 6 Device Drivers */
+    if(_IOC_TYPE(cmd) != AESD_IOC_MAGIC)
+        return -ENOTTY;
+    if(_IOC_NR(cmd) > AESDCHAR_IOC_MAXNR)
+        return -ENOTTY;
+
+    
+    switch (cmd)
+    {
+        case AESDCHAR_IOCSEEKTO:
+            if(!(copy_from_user(&buffer, (const void __user *)arg, sizeof(buffer))))
+            {
+                if (buffer.write_cmd >= AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED)
+                {
+                    ret = -EINVAL;
+                    return ret;
+                }
+                if (buffer.write_cmd_offset > dev->cb.entry[buffer.write_cmd].size)
```

Comment:
@malolasimman this state could change between the check here and the locking interval below. It would be best to have this access inside the locking interval.